### ISSUE TITLE
Chrome headless 500 Internal Server Error 

### ISSUE THREAD

**Issue by juban** [drud/ddev-contrib#192]
Hi,

I'm trying to setup a headless chrome as a service following the headless-chrome service (for Behat Testing) and have the following issue:

Service is running fine, and I can access http://localhost:9222 alright from the host.
However, when running `wget http://chrome:9222` from the web container, i constantly get a `500 Internal Server Error` response.
I'm running lastest ddev 1.18.2 and Docker Desktop 4.4.2 on a M1 MacBook Pro.

Does anyone else is experiencing this issue?

**Comment by rfay** [comment]
I assume you're using https://github.com/drud/ddev-contrib/tree/master/docker-compose-services/headless-chrome

It sure sounds like you are running the amd64 version of the chrome image.

Please `docker pull isholgueras/chrome-headless:latest` to make sure you have the right one. 

Then after `ddev restart`, `docker exec -s chrome arch`  and see what it says, amd64 or arm64. 

I'm not sure wget is the right tool to be using there inside the web container though, I'd use `curl -I http://chrome:9222` to start analyzing the connection.

**Comment by juban** [comment]
Thank you @rfay 

Yes, I'm pretty sure I'm using latest image build here (did a pull as suggested anyway to be double sure).

When checking the arch value, i'm getting `aarch64` which is `arm` if I'm not wrong, so should be good.

However, trying to curl the container on port 9222 still gives different results depending on if the request is sent from the host or from within the web container:

From host:
```
curl -I http://localhost:9222
HTTP/1.1 200 OK
Content-Length:1540
Content-Type:text/html; charset=UTF-8
```

From web container:
```
curl -I http://chrome:9222
HTTP/1.1 500 Internal Server Error
Content-Length:63
Content-Type:text/html
```

So, same results using wget, curl or whatever utility allowing to send a web request.

This is odd.

**Comment by jonaseberle** [comment]
The body of that response contains an error message:
`Host header is specified and is not an IP address or localhost.`
wget/curl seem to transmit `Host: chrome:9222` (see `curl -v`) and chrome does not seem to like that.

This does not return an error: 
`curl -v -H 'Host:' chrome http://chrome:9222`

**Comment by juban** [comment]
After double thoughts, the `aarch64` arch value looks strange. The image is available for `arm64` or `amd64`. How come I end up with a `aarch64` version? Could it be related to a docker-compose issue maybe?

**Comment by juban** [comment]
@jonaseberle Thank you for your help.

Indeed, accessing the chrome instance using the internal docker IP of the chrome container works fine.
I guess that Behat manages to send the additional host header which make it work in that context.
However, I'm trying to do the same using a `puppeteer.connect()` and I can't figure out how I could send that kind of information. I could set the `browserURL` to the dynamic internal IP, but it's not very convenient in the end.

@rfay Any idea? (Sorry, I know it's probably not directly a ddev issue though)

**Comment by juban** [comment]
OK, I managed to do this by using the `dns` Nodejs lib. Something like:

```
const puppeteer = require('puppeteer');
let chromeIP = "";
const dns = require("dns");
dns.lookup('chrome', function(err, result) {
    chromeIP = result;
});
const browserPromise = puppeteer.connect({
    browserURL: `http://${chromeIP}:9222`
})
```

Not very straightforward, but it works.

**Comment by rfay** [comment]
aarch64 is fine. That's what linux reports for arm64.

**Comment by rfay** [comment]
I don't think this should be closed unless regular functionality of https://github.com/drud/ddev-contrib/tree/master/docker-compose-services/headless-chrome is confirmed... @juban were you actually doing something out of the ordinary? @isholgueras had tested this and determined all was fine.

**Comment by juban** [comment]
Not really :)

Some context here to help you understand what I'm trying to acheive (I'm not there yet ^^):

We are running a vitejs configuration within the ddev web container. This works fine on Linux and Mac Intel configs, but things get a bit complicated on M1 because some of the vitejs dependencies (webpack is also affected) are not available for arm64 yet. For instance `rollup-plugin-imagemin` can't be installed for now (but that's another story). We are also using `rollup-plugin-critical` which depends on puppeteer to extract critical css using a chrome instance. Sadely, for now, even when running the web container using rosetta2 (custom Dockerfile by adding `FROM --platform=linux/amd64 $BASE_IMAGE`), the provided chromium binary won't run, not matter what. So, I'm trying to use a chrome instance privided as a Docker service and try to connect puppeteer to it. Using the trick above (the dns Nodejs lib), I'm finaly able to connect to the service. I still get an error though, related to puppeteer unable to access the temp file, but that's an unrelated issue I think.

Anyway, being able to use chrome as a service will also be useful in order to run acceptance tests with codeception for instance (I did not try that yet on M1).

**Comment by rfay** [comment]
That sounds out of the ordinary to me!  :)

The recommended thing, of course, is to do everything you can to get the upstream components provided for arm64. I'm sure you know that the reason the headless-chrome image had to change to chromium was that chrome isn't available for arm64.

**Comment by juban** [comment]
Haha, agreed, maybe not so ordinary :)

Anyway, I might have lost you on the rosetta2 explanation thing which is unrelated to the current issue.
I'm totally aware that in order to have an chrome instance running, I need to use the extra arm64 optimized container, and that's the all point. I want to be able to communicate with that container from the web container to run whatever necessary process.
That said, if the Behat implementation runs fine on a M1 config, I guess we can close the ticket (that's ultimately more a isholgueras/chrome-headless image related issue I think).

**Comment by juban** [comment]
For whoever interested in that issue, I found the following workaround to use puppeteer on M1 :
In ddev config, in the `web_extra` property, add `chromium` as dependancy and add `PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium` in `web_environment` property.
So basicaly, not rely on the external chrome service at all.

**Comment by rfay** [comment]
Could you please give the full configuration? Is this a change to the puppetteer recipe here, so should it be a PR against that?

**Comment by juban** [comment]
@rfay Yes, indeed. The recipe is not working anymore with the latest ddev web container image nor with M1 hosts.
I will try to PR a fix as soon as I can.

**Comment by juban** [comment]
Bad news: the chromium package works fine with puppeteer on M1 but process gets indefinitely stuck on intel architecture... it's driving me crazy. I guess I will have to dynamically set the usage of bundled chrome version on intel and use chromium on M1.

**Comment by rfay** [comment]
I imagine there will be some work required to understand the chromium image/package better. Thanks for your work on this!

**Comment by rfay** [comment]
Could you please say what images you're talking about explicitly? Give a link to the dockerhub images, thanks.

**Comment by juban** [comment]
I'm not using any image per say. Only using the apt chromium package on the base ddev 1.18.2 web container image.
I've seen reports of stuck processes on many puppeteer issues, but not which make much sens here. Can't understand why this is working so smoothly on M1 and chromium connection gets stuck on intel...

**Comment by rfay** [comment]
OK, interesting. That may be a better technique (after you solve this) than the separate container.

**Comment by juban** [comment]
@rfay Quick question: in your opinion, what would be the best approach to deal with conditional environment variables? Basically, I would want to only set `PUPPETEER_EXECUTABLE_PATH` if the container runs on `aarch64`.

**Comment by rfay** [comment]
So you want to set PUPPETEER_EXECUTABLE_PATH in which container? I assume you're currently thinking about setting it externally, not making any changes to the image. You can certainly change the image to choose what to launch.

**Comment by juban** [comment]
I want to set PUPPETEER_EXECUTABLE_PATH in the web container context but only if the architecture is M1 (aarch64). To be clear, simply to be able to use either the chromium package for M1 arch and the bundled chrome binary provided by puppeteer on other architectures.

**Comment by rfay** [comment]
One super easy way then is to just set `web_environment: [PUPPETEER_EXECUTABLE_PATH=...]` in config.yaml, then override it with a `config.m1.yaml` with alternate path. 

Another easy way to do it is to launch it with a script that you install in a .ddev/web-build/Dockerfile (instead of relying on an env var). The script can check arch before launching.

**Comment by juban** [comment]
@rfay Thank you for your helpful suggestions. I've made a PR to update the current "Puppeteer Headless Chrome support" recipe to implement the fix for ddev 1.6 and upper as well as a workaround for Apple M1 arch.
